File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/AssertWellFormedBreakTargets.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {CompilerError} from '..';
import {BlockId, ReactiveFunction, ReactiveTerminalStatement} from '../HIR';
import {ReactiveFunctionVisitor, visitReactiveFunction} from './visitors';

/**
 * Assert that all break/continue targets reference existent labels.
 */
export function assertWellFormedBreakTargets(fn: ReactiveFunction): void {
  visitReactiveFunction(fn, new Visitor(), new Set());
}

class Visitor extends ReactiveFunctionVisitor<Set<BlockId>> {
  override visitTerminal(
    stmt: ReactiveTerminalStatement,
    seenLabels: Set<BlockId>,
  ): void {
    if (stmt.label != null) {
      seenLabels.add(stmt.label.id);
    }
    const terminal = stmt.terminal;
    if (terminal.kind === 'break' || terminal.kind === 'continue') {
      CompilerError.invariant(seenLabels.has(terminal.target), {
        reason: 'Unexpected break to invalid label',
        loc: stmt.terminal.loc,
      });
    }
  }
}

File: compiler/crates/react_compiler_reactive_scopes/src/assert_well_formed_break_targets.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! Assert that all break/continue targets reference existent labels.
//!
//! Corresponds to `src/ReactiveScopes/AssertWellFormedBreakTargets.ts`.

use std::collections::HashSet;

use react_compiler_hir::{
    BlockId, ReactiveFunction, ReactiveTerminal, ReactiveTerminalStatement,
    environment::Environment,
};

use crate::visitors::{visit_reactive_function, ReactiveFunctionVisitor};

/// Assert that all break/continue targets reference existent labels.
pub fn assert_well_formed_break_targets(func: &ReactiveFunction, env: &Environment) {
    let visitor = Visitor { env };
    let mut state: HashSet<BlockId> = HashSet::new();
    visit_reactive_function(func, &visitor, &mut state);
}

struct Visitor<'a> {
    env: &'a Environment,
}

impl<'a> ReactiveFunctionVisitor for Visitor<'a> {
    type State = HashSet<BlockId>;

    fn env(&self) -> &Environment { self.env }

    fn visit_terminal(
        &self,
        stmt: &ReactiveTerminalStatement,
        seen_labels: &mut HashSet<BlockId>,
    ) {
        if let Some(label) = &stmt.label {
            seen_labels.insert(label.id);
        }
        let terminal = &stmt.terminal;
        match terminal {
            ReactiveTerminal::Break { target, .. } | ReactiveTerminal::Continue { target, .. } => {
                assert!(
                    seen_labels.contains(target),
                    "Unexpected break/continue to invalid label: {:?}",
                    target
                );
            }
            _ => {}
        }
        // Note: intentionally NOT calling self.traverse_terminal() here,
        // matching TS behavior where visitTerminal override does not call traverseTerminal.
        // Recursion into child blocks happens via traverseBlock→visitTerminal for nested blocks.
        // The TS visitor only checks break/continue at the block level, not terminal child blocks.
    }
}
